controllers/apply/form-router-next/steps.js

```diff
@@ -132,23 +137,90 @@ module.exports = function(formBuilder) {
 
             const stepIndex = parseInt(req.params.step, 10) - 1;
             const step = currentSection.steps[stepIndex];
+            const stepFields = flatMap(step.fieldsets, 'fields');
+
+            // Check if this step expected any file inputs
+            const fileFields = stepFields
+                .filter(f => f.type === 'file')
+                .map(f => f.name);
+
+            // If we expected files, check whether they were sent
+            const filesToUpload = fileFields
+                .filter(fieldName => {
+                    // Retrieve the file from Formidable's parsed data
+                    const uploadedFile = get(req.files, fieldName);
+                    // Ensure a file was actually provided
+                    // (eg. ignore empty file inputs when a file already exists)
+                    return uploadedFile && uploadedFile.size > 0;
+                })
+                .map(fieldName => {
+                    // Log this file as a potential upload
+                    const uploadedFile = get(req.files, fieldName);
+                    return {
+                        fieldName: fieldName,
+                        fileData: uploadedFile
+                    };
+                });
+
+            // Append the file data to the overall form data for validation
+            filesToUpload.forEach(file => {
+                data[file.fieldName] = {
+                    filename: file.fileData.name,
+                    size: file.fileData.size,
+                    type: file.fileData.type
+                };
+            });
 
-            const validationResult = validateForm(form, data);
+            let validationResult = validateForm(form, data);
 
             try {
-                await PendingApplication.saveApplicationState(
-                    currentlyEditingId,
-                    validationResult.value
-                );
-
                 const fieldNamesForStep = flatMap(step.fieldsets, 'fields').map(
                     field => field.name
                 );
 
-                const errorsForStep = validationResult.messages.filter(item =>
+                let errorsForStep = validationResult.messages.filter(item =>
                     fieldNamesForStep.includes(item.param)
                 );
 
+                const [invalidFiles, validFiles] = partition(
+                    filesToUpload,
+                    file => {
+                        return errorsForStep.find(
+                            _ => _.param === file.fieldName
+                        );
+                    }
+                );
+
+                invalidFiles.forEach(file => {
```

Comment:
I reckon we could avoid mutating the value here using `pickBy`

```js
const newData = pickBy(validationResult.value, field => invalidFiles.map(field => file.fieldName).includes(filed.name) === false)
```

Or something similar to that